Support image upload and storing using Blobstore

anna-giang opened the pull request:
Add a form field to allow user to upload an optional image with their comment.
Store this image using Blobstore.
Delete the images Blobs when the user deletes all comments.

anna-giang committed:
Rename comment form div ID

Rename the div containing comment form (wrapper div for styling purposes): 'comment-form' --> 'comment-form-div'.
Add id='somment-form' to the form element representing the comment form.
This is to prepare for mapping the blobstore upload url to the form action (need to be able to identify the form by an id).

anna-giang committed:
Add image upload field in comment form and fix style

Fix the spacing between form fields.
Blobstore and other backend functionality will be linked to this form in future.

anna-giang committed:
Add servlet that returns the blobstore upload url

anna-giang committed:
Fetch blobstore upload url from servlet on page load and add to form

Call getBlobstoreUrl() onload to fetch the blobstore upload url from the BlobstoreUploadUrlServlet, and add as the 'action' of the comment-form.

anna-giang committed:
Retrieve url of  uploaded image from BlobstoreService and store in datastore

When user submits form, the form will send inputted data to the URL returned by the BlobstoreUploadUrlServlet, where the image is processed and stored.
The BlobstoreUploadUrlServlet will then redirect to the DataServlet at '/data' endpoint where the image url is then retreived from the BlobstoreService and stored in datastore as part of the Comment entity.
Also added encoding type in <form> element in index.html - necessary for successful storage of the image.

anna-giang committed:
Update class header comment to describe class function

anna-giang committed:
Send uploaded image from backend to frontend and display

Add uploaded image url to the JSON object returned on GET from DataServlet, and display the image with associated comment. Images are also linked to their respective urls.

anna-giang committed:
Fix style of comments to fit in uploaded images

Edit script.js to add <div>s  to wrap around each of comment text and comment image elements when generating the comment output. Better divides each comment into separate sections for CSS styling and layout on the page.

anna-giang committed:
Delete stored images when comment is deleted

Add backend support to delete Blob storing the image when the associated comment are deleted.
Previously, all comment entities were deleted but the actual images would still be stored somewhere using BlobStore.
Now, all images are deleted along with their BlobInfo metadata.

thanhdng commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -67,6 +81,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
```
nit: remove this line.

anna-giang replied:
done!

thanhdng commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -111,4 +137,55 @@ private String convertToJson(List<HashMap<String, String>> list) {
+   * @return URL of the uploaded, or none if no image was uploaded
+   */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName, BlobKey blobKey) {
+    
```
nit: remove this line

anna-giang replied:
done!

thanhdng commented on portfolio/src/main/webapp/script.js:
```diff
@@ -77,15 +77,25 @@ function fetchComments(quantity=5) {
+        
+        // Display image as well, if there is an image
+        let imageUrl = commentData[i].attachedImage;
+        if (imageUrl == null) {
```
|commentContent| share the same first part (|commentAuthor| and |commentText|) with or without |imageUrl|. Could you take that into account and make this part cleaner?

anna-giang replied:
I have two different solutions for this one - which one is better practice do you think?

I could do it like this, just moving the shared section of the string outside of the if-block:

```
// Display image as well, if there is an image
let imageUrl = commentData[i].attachedImage;
commentContent += '<div class="comment"><div class="flex-item"><p class="body-text"><b>' + commentAuthor + '</b></p>' 
    + '<p class="body-text">' + commentText + '</p></div>'; // outer <div> not closed yet

if (imageUrl == null) {
  commentContent += '</div>'; // close outer div
}
else {
  commentContent += '<div class="flex-item"><a href=' + imageUrl + ' target="_blank"><img class="comment-image" src=' + imageUrl + '></a></div></div>'; // add image
}
```

or I could do it like this, using `document.createElement` to generate all the elements instead of writing HTML in strings:

There is a lot more lines of code, but it might make it easier to extend the comments display later on.

```
function fetchComments(quantity=5) {
  
  const endpoint = '/data?';
  var queryString = new URLSearchParams();
  queryString.append('quantity', String(quantity));
  const url = endpoint + queryString.toString();

  fetch(url).then(response => response.json()).then((commentData) => {
    
    // Delete what is currently displayed
    document.getElementById('comment-list').innerHTML = "";
    
    if (commentData.length === 0 ) { // if there are no existing comments
      var commentDiv = document.createElement('div'); // div to store one comment
      commentDiv.className = "comment";
      
      // disable the delete comments button & display "No Comments"
      document.getElementById('delete-comments-button').setAttribute('disabled','true');
      let content = document.createElement('p');
      content.className = "body-text";
      content.innerText = "No Comments";
      commentDiv.appendChild(content);

      document.getElementById('comment-list').appendChild(commentDiv);

    }
    else { // otherwise display the comments and any attached image
      
      for (let i = 0; i < commentData.length; i++) {
        
        var commentDiv = document.createElement('div'); // div to store one comment
        commentDiv.className = "comment";

        let commentText = commentData[i].commentText; 
        // Comment author name: "'Anonymous' if there was not a name submitted."
        let commentAuthor = commentData[i].commentAuthor === "" ? "Anonymous" : commentData[i].commentAuthor; 
        let imageUrl = commentData[i].attachedImage;

        let textContainer = document.createElement('div'); // Container for comment text & author name
        textContainer.className = "flex-item";
        
        // Author name <p> element
        let authorP = document.createElement('p');
        authorP.className = "body-text";
        let bolding = document.createElement('b');
        bolding.innerText = commentAuthor;
        authorP.appendChild(bolding);

        // Comment text <p> element
        let commentP = document.createElement('p');
        commentP.className = "body-text";
        commentP.innerText = commentText;

        textContainer.appendChild(authorP);
        textContainer.appendChild(commentP);
        commentDiv.appendChild(textContainer);

        // Display image as well, if there is an image
        if (imageUrl) {
          let imageContainer = document.createElement('div');
          imageContainer.className = "flex-item";
          
          let imageLink = document.createElement('a');
          imageLink.href = imageUrl;
          imageLink.target = "_blank"; // link will open in new tab
          
          let image = document.createElement('img');
          image.className = "comment-image";
          image.src = imageUrl;

          imageLink.appendChild(image);
          imageContainer.appendChild(imageLink);
          commentDiv.appendChild(imageContainer);
        }
        document.getElementById('comment-list').appendChild(commentDiv);
      }
    }
  });
}
```

thanhdng replied:
I would go with the 1st option. 2nd option is nice, but is a bit overkill imo.

anna-giang replied:
Yep, changed it to the 1st option! :)

thanhdng reviewed (commented):
Please see comments inline :)

anna-giang committed:
remove unnecessary blank lines

anna-giang committed:
Refactor displaying of comment images to reduce repeated code

Move the shared section of the HTML string representing a comment to outside of the if-block.

thanhdng reviewed (approved):
LGTM
